changefeedccl: block db-level changefeed creation for non-empty tableset

log-head opened the pull request:
On startup, the changefeed should not be started until there exist tables in the database to watch. If no tables exist at the time of startup, then a tableset watcher is created and polled; the changefeed will be started from where there is first a target table.

Fixes: #148833
Epic: CRDB-1421

Release note: none

log-head committed:
changefeedccl: block db-level changefeed creation for non-empty tableset

On startup, the changefeed should not be started until there exist
tables in the database to watch. If no tables exist at the time of
startup, then a tableset watcher is created and polled; the changefeed
will be started from where there is first a target table.

Fixes: #147371
Epic: CRDB-1421

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154163)
<!-- Reviewable:end -->

log-head commented:
This works if there's sufficient time between creating the db-level changefeed and going from 0->1 tables in the database. Otherwise, through planning the changefeed, the table descriptors are being fetched at times where they don't exist. I'm still fiddling with a follow-up commit to fix the timestamp where table descriptors are queried.

log-head committed:
changefeedccl: Fix timestamp for descriptor fetches during changefeed creation

When a changefeed is planned and executed, there are several places
where target table descriptors are fetched. With a db-level changefeed,
the set of tables can change during changefeed startup. Now, the
timestamp is set to the initial high water for retrieving table
descriptors.  Fixes: #154549 Epic: CRDB-1421

Release note: None
